### PR TITLE
Refactor multi target search

### PR DESCRIPTION
**Pull request by dirkkul**
### What's being changed:

Refactors multi-target search to be more effizient in multi node environments:
- concurrency (different target vectors) is now done at the shard level. A shard always contains all targetvectors for an object.
- computation of missing distances and combinationof different target vectors are also done on the shard level. No calls to other nodes are needed, simplifying the whole setup 

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [x] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/index.go [hunk 1]
```diff
@@ -1748,7 +1682,7 @@ func (i *Index) IncomingSearch(ctx context.Context, shardName string,
 		return nil, nil, enterrors.NewErrUnprocessable(fmt.Errorf("local %s shard is not ready", shardName))
 	}
 
-	if searchVector == nil {
+	if searchVectors == nil {
```
Comment (antas-marcin) [inline adapters/repos/db/index.go:1685]: hmm maybe `if len(searchVectors) == 0` check would be better here?

File: adapters/repos/db/shard_combine_multi_target.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+		f := func() error {
+			distances, err := shard.VectorDistanceForQuery(ctx, allIDs[id].ID(), targets.searchVector, targets.target)
+			mutex.Lock()
+			if err != nil {
+				// when we cannot look up missing distances for an object, it will be removed from the result list
+				combinedResults.RemoveIdFromResult(id)
+			} else {
+				combinedResults.AddScores(id, targets.target, distances, weights)
+			}
+			mutex.Unlock()
```
Comment (antas-marcin) [inline adapters/repos/db/shard_combine_multi_target.go:269]: maybe safe would be to put the `mutex.Unlock()` to line 263 with `defer mutex.Unlock()`? we are sure that nothing can go wrong during `combinedResults.RemoveIdFromResult` or `combinedResults.AddScores`?

File: modules/multi2vec-dummy/clients/meta.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package clients
+
+type MetaClient struct{}
+
+func (v MetaClient) MetaInfo() (map[string]interface{}, error) {
```
Comment (antas-marcin) [inline modules/multi2vec-dummy/clients/meta.go:16]: is this `multi2vec-dummy` module finished? It looks like a half-done implementation

Comment (dirkkul) [reply modules/multi2vec-dummy/clients/meta.go]: woops, thanks for noticing!

File: adapters/repos/db/shard_combine_multi_target.go (hunk 2 above)
Comment (dirkkul) [reply adapters/repos/db/shard_combine_multi_target.go:269]: they are very simple, but using defer doesn't hurt

**Review by antas-marcin — APPROVED** [review]

**Pull request merged**